enable degree program filtering dropdown

Urjikoo opened the pull request:
Created a degree program dropdown with Both,Undergraduate and Graduate options to filter the courses. When the degree program is selected from the dropdown, the unselected courses will fade out and will not be clickable for the pop over.
## Types of changes
<!--
  What types of changes does your code introduce? Put an `x` in all the boxes
  that apply:
-->
- [ ]  Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality
      to change)

## Checklist:
<!--
  Go over all the following points, and put an `x` in all the boxes that apply.
-->
- [x] I have run `eslint` on the code
- [x] I have added JSDoc for all of my code (where applicable)
- [x] I have added tests to cover my changes.

## Priority:
- [x] Normal <!-- New piece of functionality -->
- [ ] High <!-- Critical bug requiring urgent review -->

## Related Issues:
<!--
  Add the issue number in below that this PR is intended to address.

  Also mention any issues that this PR is related to, and if possible,
  provide more information regarding the relationship to the issues in
  question (i.e does the PR fix the issue, is it designed to fix another bug
  that is similar to the issue etc.)
-->
Fixes #558 

<!--
  Attribution:
  PR Template adapted from: https://github.com/h5bp/html5-boilerplate/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

Urjikoo committed:
enable degree program filtering dropdown

Urjikoo committed:
enable degree program filtering dropdown

Urjikoo committed:
enable degree program filtering dropdown

Urjikoo committed:
lint test passed and restored aria-disabledattr

Urjikoo committed:
fixed the fomatting and indentation

In this commit, I updated the formatting and indentation accoding to the existing code and removed unused prop along with comments thats no longer needed. I also added comment for the degree program enum.

Urjikoo committed:
Merge branch 'develop' into feature/558-enable-filtering-degree-program-dropdown

Urjikoo committed:
Removed isFaded prop as it's not being used anymore.
 The disabled prop has replaced isFaded prop, therefore isFaded is removed.

natalynyu reviewed (changes requested):
Thanks for making this PR, Urji! We are on the right track, and the filter is working as expected. Some of my comments are related to styling and future-proofing the code, and the others are related to testing. I plan to go over these comments together so that you can get further clarification if needed, but feel free to reach out with any other questions too.

Urjikoo committed:
indentation updated and dropdown test modified.
updated dropdown test to test whether the courses are disabled or not based degree program selections. The opacity for disabled courses is also updated to markone theme. Indentaion and spaces between lines have been updated.

Urjikoo commented on src/client/components/pages/Schedule/__tests__/SchedulePage.test.tsx:
```diff
@@ -88,6 +89,75 @@ describe('Schedule Page', function () {
+        { target: { value: DEGREE_PROGRAM.BOTH } });
+      const undergradCourseId = testCourseScheduleData[0].courses[0].id;
+      const undergradFaded = (renderResult.getByTestId(undergradCourseId).getAttribute('data-disabled') === 'true');
+      strictEqual(selectedDegreeProgram === 'BOTH', undergradFaded === false);
```
The test now updated with looping through testCourseScheduleData and grabbing the courses and coursesprefix.

Urjikoo commented on src/client/components/pages/Schedule/blocks/CourseListingButton.tsx:
```diff
@@ -25,7 +18,7 @@ const CourseListingButton = styled.button.attrs({
   border: none;
   font-size: inherit;
-  opacity: ${({ isFaded }) => (isFaded ? '0.5' : '1')};
+  opacity: ${({ disabled }) => (disabled ? '0.5' : '1')};
```
This was great suggestion! imported `fromTheme` from 'mark-one' and changed the opacity and color for coursePrefix buttons.

natalynyu replied:
Nice! Thanks for updating the cursor as well.

Urjikoo commented on src/client/components/pages/Schedule/SchedulePage.tsx:
```diff
@@ -167,6 +175,23 @@ const SchedulePage: FunctionComponent = () => {
+          options={[{ label: 'Both', value: DEGREE_PROGRAM.BOTH },
+            { label: 'Graduate', value: DEGREE_PROGRAM.GRADUATE },
+            { label: 'Undergraduate', value: DEGREE_PROGRAM.UNDERGRADUATE }]}
+          value={selectedDegreeProgram.toString()}
```
Removed `toString()` as the value is already a string.

Urjikoo commented on src/client/components/pages/Schedule/SchedulePage.tsx:
```diff
@@ -167,6 +175,23 @@ const SchedulePage: FunctionComponent = () => {
+          isLabelVisible
+          options={[{ label: 'Both', value: DEGREE_PROGRAM.BOTH },
+            { label: 'Graduate', value: DEGREE_PROGRAM.GRADUATE },
+            { label: 'Undergraduate', value: DEGREE_PROGRAM.UNDERGRADUATE }]}
```
Updated both enum file and also options block.

Urjikoo commented on src/client/components/pages/Schedule/__tests__/SchedulePage.test.tsx:
```diff
@@ -88,6 +89,75 @@ describe('Schedule Page', function () {
+    afterEach(function () {
+      clock.uninstall();
+    });
+    it('renders the list of degree program options ', function () {
```
the extra space is removed.

Urjikoo commented on src/client/components/pages/Schedule/__tests__/SchedulePage.test.tsx:
```diff
@@ -88,6 +89,75 @@ describe('Schedule Page', function () {
       strictEqual(term, TERM.FALL);
     });
   });
+  describe('Degree Program Dropdown', function () {
```
The it blocks are now organized and no longer in separate describe block.

Urjikoo commented on src/client/components/pages/Schedule/__tests__/SchedulePage.test.tsx:
```diff
@@ -88,6 +89,75 @@ describe('Schedule Page', function () {
+        { target: { value: DEGREE_PROGRAM.GRADUATE } });
+      const gradCourseId = testCourseScheduleData[0].courses[0].id;
+      const gradFaded = (renderResult.getByTestId(gradCourseId).getAttribute('data-disabled') === 'true');
+      strictEqual(gradCourseId === 'Graduate', gradFaded === false);
```
The test is now updated with `strictEqual(targetCourseListing.hasAttribute('disabled'), false, `${course.coursePrefix} ${course.courseNumber}-was disabled when it shouldn't be`);
`

natalynyu replied:
The tests look great! Thanks for updating them

natalynyu commented on src/client/components/pages/Schedule/__tests__/SchedulePage.test.tsx:
```diff
@@ -88,6 +89,110 @@ describe('Schedule Page', function () {
+          const targetCourseListing = within(
+            targetBlock.parentElement
+          ).getByText(course.courseNumber);
+          strictEqual(targetCourseListing.hasAttribute('disabled'), course.isUndergraduate, `${course.coursePrefix} ${course.courseNumber}-was disabled when it shouldn't be`);
```
I would check that the course is not a graduate course rather than checking that the course is undergraduate to future proof this test in case we have more categorizations of courses (beyond undergraduate & graduate) in the future.

Urjikoo replied:
Hi in the data we have now, the courses only has one prop `isUndergraduate: true` or `isUndergraduate: false` unless we create an array for graduate courses in test. should i loop through the data and create grad course array?

natalynyu replied:
Yes I think that would make sense here.

natalynyu reviewed (changes requested):
This is looking great! Thanks for adding in the change to the cursor styles - I had forgotten about that. I just had a few remaining minor comments.

Urjikoo committed:
space has been added
space has been added after the semi-colon courseId and isUndergraduate variables.

Urjikoo committed:
context has been removed
i removed the context as we no longer need it because the describe gives enough description.

Urjikoo commented on src/client/components/pages/Schedule/__tests__/SchedulePage.test.tsx:
```diff
@@ -88,6 +89,110 @@ describe('Schedule Page', function () {
+          const targetCourseListing = within(
+            targetBlock.parentElement
+          ).getByText(course.courseNumber);
+          strictEqual(targetCourseListing.hasAttribute('disabled'), false, `${course.coursePrefix} ${course.courseNumber}-was disabled when it shouldt be`);
```
Typo has been fixed.

Urjikoo commented on src/client/components/pages/Schedule/__tests__/SchedulePage.test.tsx:
```diff
@@ -88,6 +89,110 @@ describe('Schedule Page', function () {
+      isUndergraduate:boolean,
+      coursePrefix: string,
+      courseNumber: string
+    }[];
```
the space has been added for `courseId` and `isUndergraduate`.

Urjikoo commented on src/client/components/pages/Schedule/__tests__/SchedulePage.test.tsx:
```diff
@@ -88,6 +89,110 @@ describe('Schedule Page', function () {
+    afterEach(function () {
+      clock.uninstall();
+    });
+    context('the degree dropdown is selected', function () {
```
i removed the context as the describe gives enough information.

natalynyu reviewed (approved):
This looks great! Feel free to merge when ready.
